Xamarin.Forms.Platform.Android/AppCompat/TabbedPageRenderer.cs

```diff
@@ -407,55 +632,329 @@ void UpdateTabIcons()
-				if (tintColor.IsDefault && _backgroundDrawable != null)
-					_tabLayout.SetBackground(_backgroundDrawable);
-				else if (!tintColor.IsDefault)
+				if (_originalTabTextColors == null)
+					_originalTabTextColors = _tabLayout.TabTextColors;
+			}
+
+			Color barItemColor = BarItemColor;
+			Color barTextColor = Element.BarTextColor;
+			Color barSelectedItemColor = BarSelectedItemColor;
+
+			if (barItemColor.IsDefault && barTextColor.IsDefault && barSelectedItemColor.IsDefault)
+				return _originalTabTextColors;
+
+			if (_newTabTextColors != null)
+				return _newTabTextColors;
+
+			int checkedColor;
+			int defaultColor;
+
+			if (!barTextColor.IsDefault)
+			{
+				checkedColor = barTextColor.ToAndroid().ToArgb();
+				defaultColor = checkedColor;
+			}
+			else
+			{
+				defaultColor = barItemColor.ToAndroid().ToArgb();
+
+				if (barItemColor.IsDefault && _originalTabTextColors != null)
+				{
+					defaultColor = _originalTabTextColors.DefaultColor;
+				}
+
+				checkedColor = defaultColor;
+
+				if (!barSelectedItemColor.IsDefault)
 				{
-					if (_backgroundDrawable == null)
-						_backgroundDrawable = _tabLayout.Background;
-					_tabLayout.SetBackgroundColor(tintColor.ToAndroid());
+					checkedColor = barSelectedItemColor.ToAndroid().ToArgb();
+				}
+			}
+
+			_newTabTextColors = GetColorStateList(defaultColor, checkedColor);
+			return _newTabTextColors;
+		}
+
+		protected virtual ColorStateList GetItemIconTintColorState()
+		{
+			if (IsDisposed)
+				return null;
+
+			if (IsBottomTabPlacement)
+			{
+				if (_orignalTabIconColors == null)
+					_orignalTabIconColors = _bottomNavigationView.ItemIconTintList;
+
+			}
+			// this ensures that existing behavior doesn't change
+			else if (!IsBottomTabPlacement && BarSelectedItemColor.IsDefault && BarItemColor.IsDefault)
+			{
+				return null;
+			}
+
+			Color barItemColor = BarItemColor;
+			Color barSelectedItemColor = BarSelectedItemColor;
+
+			if (barItemColor.IsDefault && barSelectedItemColor.IsDefault)
+			{
+				return _orignalTabIconColors;
+			}
+
+			if (_newTabIconColors != null)
+				return _newTabIconColors;
+
+			int defaultColor = barItemColor.ToAndroid().ToArgb();
+
+			if (barItemColor.IsDefault && _orignalTabIconColors != null)
+			{
+				defaultColor = _orignalTabIconColors.DefaultColor;
+			}
+
+			int checkedColor = defaultColor;
+
+			if (!barSelectedItemColor.IsDefault)
+			{
+				checkedColor = barSelectedItemColor.ToAndroid().ToArgb();
+			}
+
+			_newTabIconColors = GetColorStateList(defaultColor, checkedColor);
+			return _newTabIconColors;
+		}
+
+		public bool OnNavigationItemSelected(IMenuItem item)
+		{
+			if (Element == null || IsDisposed)
+				return false;
+
+			int selectedIndex = item.Order;
+			if (_bottomNavigationView.SelectedItemId != item.ItemId && Element.Children.Count > selectedIndex && selectedIndex >= 0)
+			{
+				Element.CurrentPage = Element.Children[selectedIndex];
+			}
+
+			return true;
+		}
+
+		bool IsDisposed
```

Comment:
why do we need this?

Reply:
The naming might be confusing but I was just consolidating these checks into a single place. Which got a bit verbose with the addition of the _bottomNavigation and _relativelayout checks

https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.Android/AppCompat/TabbedPageRenderer.cs#L414
https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.Android/AppCompat/TabbedPageRenderer.cs#L444

Which I'm guessing exists due to race conditions